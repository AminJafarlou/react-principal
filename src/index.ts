export * from "./createStore";
export * from "./reducer";
export * from "./provider";
export * from "./persist";
