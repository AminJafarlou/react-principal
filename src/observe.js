// @ts-check
import React from "react";
import invariant from "invariant";
import { __DEV__ } from "./utils";

function useContextWithObserve(context, nextObserveState) {
  const stateKeys = getKeys(context._currentValue);

  let observeKeys = stateKeys;

  if (nextObserveState) {
    if (__DEV__) {
      invariant(
        Array.isArray(nextObserveState),
        `nextObserveState expected to be an Array of string but ${typeof nextObserveState} was received`,
      );
    }
    observeKeys = nextObserveState;
  }

  const observeBit = getBits(stateKeys, observeKeys);

  return readContext(context, observeBit);
}

function getKeys(obj) {
  return Object.keys(obj).sort();
}

function calculateChangedBits(prev, next) {
  let changed = {};
  const keys = getKeys(next);

  if (__DEV__) {
    invariant(
      keys.length < 32,
      `expected return state have property least than 32 but has ${keys.length} property`,
    );
  }

  keys.forEach((key) => {
    if (prev[key] !== next[key]) {
      changed[key] = next[key];
    }
  });

  const result = getBits(keys, changed);

  return result;
}

function getBits(keys, usage) {
  let result = 0;
  keys.forEach((key, index) => {
    if (usage.hasOwnProperty(key)) {
      // eslint-disable-next-line no-bitwise
      result |= 1 << index % 31;
    }
  });
  return result;
}

const ReactCurrentDispatcher =
  // @ts-ignore
  React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
    .ReactCurrentDispatcher;

function readContext(Context, observedBits) {
  const dispatcher = ReactCurrentDispatcher.current;
  if (dispatcher === null) {
    throw new Error(
      "react-principal: useState and useDispatch may only be called from within a " +
        "component's render. They are not supported in event handlers or " +
        "lifecycle methods.",
    );
  }
  return dispatcher.readContext(Context, observedBits);
}

export { calculateChangedBits, useContextWithObserve };
