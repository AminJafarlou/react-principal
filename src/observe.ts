import React from "react";
//@ts-ignore
import invariant from "invariant";
import { __DEV__ } from "./utils";

function useContextWithObserve(
  context: React.Context<any>,
  initialState: object,
  nextObserveState?: any[],
) {
  const stateKeys = getKeys(
    // @ts-ignore
    initialState,
  );

  // default observe to whole state
  let observeKeys = stateKeys;

  if (nextObserveState) {
    if (__DEV__) {
      invariant(
        Array.isArray(nextObserveState),
        `useState: nextObserveState expected to be an Array of string but ${typeof nextObserveState} was received`,
      );
      nextObserveState.forEach((observeKey) => {
        invariant(
          stateKeys.includes(observeKey),
          `useState: nextObserveState expected to be an Array of state keys but ${observeKey} is not one of the state`,
        );
      });
    }
    observeKeys = nextObserveState.sort();
  }

  const observeBit = getBits(stateKeys, observeKeys);

  return readContext(context, observeBit);
}

function getKeys(obj: object) {
  return Object.keys(obj).sort();
}

function calculateChangedBits(prev: { [x: string]: any }, next: any) {
  let changed: { [x: string]: any } = {};
  const keys = getKeys(next);

  if (__DEV__) {
    invariant(
      keys.length < 32,
      `expected return state have property least than 32 but has ${keys.length} property`,
    );
  }

  keys.forEach((key) => {
    if (prev[key] !== next[key]) {
      changed[key] = next[key];
    }
  });

  const changedKeys = getKeys(changed);

  const result = getBits(keys, changedKeys);

  return result;
}

function getBits(keys: string[], usage: string[]) {
  let result = 0;
  /**
   * @param {string} key
   * @param {number} index
   */
  keys.forEach((key, index) => {
    if (usage.includes(key)) {
      // eslint-disable-next-line no-bitwise
      result |= 1 << index % 31;
    }
  });

  return result;
}

const ReactCurrentDispatcher =
  // @ts-ignore
  React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
    .ReactCurrentDispatcher;

function readContext(Context: any, observedBits: number) {
  const dispatcher = ReactCurrentDispatcher.current;
  if (dispatcher === null) {
    throw new Error(
      "react-principal: useState and useDispatch may only be called from within a " +
        "component's render. They are not supported in event handlers or " +
        "lifecycle methods.",
    );
  }
  return dispatcher.readContext(Context, observedBits);
}

export { calculateChangedBits, useContextWithObserve };
