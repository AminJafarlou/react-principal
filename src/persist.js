//@ts-check
import { __DEV__ } from "./utils";

const INITIALIZE_STATE_FROM_STORAGE = Symbol();

/**
 * @param {any} Storage
 * @param {string} key
 * @param {(state: {}) => any } mapStateToPersist
 */
export const persisterCreator = function persisterCreator(
  Storage,
  key,
  mapStateToPersist,
) {
  return {
    /**
     * @param {{}} state
     * @param {{type: any}} action
     */
    persist(state, action) {
      if (action.type !== INITIALIZE_STATE_FROM_STORAGE) {
        Storage.setItem(
          key,
          JSON.stringify(mapStateToPersist ? mapStateToPersist(state) : state),
        );
      }
    },

    /**
     * @param {any} storeRef
     */
    async setToState(storeRef) {
      try {
        const storedState = await Storage.getItem(key);

        if (!storedState) {
          return;
        }

        const stateObject = JSON.parse(storedState);

        const { dispatch } = storeRef.current;
        const mappedState = mapStateToPersist
          ? mapStateToPersist(stateObject)
          : stateObject;

        dispatch({
          type: INITIALIZE_STATE_FROM_STORAGE,
          payload: mappedState,
        });
      } catch (error) {
        if (__DEV__) {
          console.error(error);
        }
      }
    },
  };
};

export const persistReducer = reducer => (state, action) => {
  if (action.type === INITIALIZE_STATE_FROM_STORAGE) {
    return { ...state, ...action.payload };
  }

  return reducer(state, action);
};
