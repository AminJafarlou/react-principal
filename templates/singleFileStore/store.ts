import { createReducer } from "react-principal";

const SET_NAME = Symbol();

interface InitialState {
  name: string;
}

const initialState: InitialState = {
  name: "",
};

const reducer = createReducer<InitialState>({
  [SET_NAME]: (state, { payload: { name } }) => ({
    ...state,
    name,
  }),
});

function setName(name: string) {
  return { type: SET_NAME, payload: { name } };
}

export { reducer, initialState, setName };
