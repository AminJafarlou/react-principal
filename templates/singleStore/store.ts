import { createReducer, createStore } from "react-principal";

const LOGIN = Symbol();
const LOGOUT = Symbol();

interface InitialState {
  accessToken?: string;
}

const initialState: InitialState = {
  accessToken: undefined,
};

const reducer = createReducer<InitialState>({
  [LOGIN]: (state, { payload: { accessToken } }) => ({
    ...state,
    accessToken,
  }),
  [LOGOUT]: () => ({
    ...initialState,
  }),
});

const AuthStore = createStore({
  reducer,
  initialState,
  persistKey: "AUTHENTICATION",
});

export { reducer, AuthStore };
