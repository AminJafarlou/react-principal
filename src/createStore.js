// @ts-check
import { createContext, useContext } from "react";
import { calculateChangedBits, useContextWithObserve } from "./observe";

/**
 * @template S
 * @typedef {object} Store
 * @property {(nextObserveState: (keyof S)[]) => S} useState
 * @property {() => (state: S) => void} useDispatch
 */

/**
 * @template S
 * @typedef {object} PrivateStore
 * @property {() => S} useState
 * @property {() => () => void} useDispatch
 * @property {React.Context<S>} stateContext
 * @property {React.Context<() => void>} dispatchContext
 * @property {() => any} reducer
 * @property {S} initialState
 */

/**
 * This function give you an store, use that in your components which want to
 * connect to store and provider
 *
 * @template T
 * @param {object} param0
 * @param {() => any} param0.reducer
 * @param {T} param0.initialState
 * @returns {Store<T>}
 */
export const createStore = ({ reducer, initialState }) => {
  const stateContext = createContext(initialState, calculateChangedBits);
  const dispatchContext = createContext(() => {});

  /** @type {PrivateStore<T>} */
  const store = {
    stateContext,
    dispatchContext,
    reducer,
    initialState,
    useState: (nextObserveState) => {
      return useContextWithObserve(
        stateContext,
        initialState,
        nextObserveState,
      );
    },
    useDispatch: () => useContext(dispatchContext),
  };

  return store;
};
