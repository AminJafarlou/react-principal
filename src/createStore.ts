import { createContext, useContext } from "react";
import { useObserveContext, createObserveContext } from "./observe";
import { INITIALIZE_STATE_FROM_STORAGE } from "./persist";
import { Reducer } from "./types";
export interface Store<S> {
  useState(nextObserveState?: (keyof S)[]): S;
  useDispatch(): (action: any, callback?: () => void) => void;
  persist(
    state: S,
    action: {
      type: any;
    },
  ): void;
  setToState(): Promise<void>;
  dispatch: (action: any, callback?: () => void) => void;
  state: S;
}

export interface PrivateStore<S> extends Store<S> {
  stateContext: React.Context<S>;
  dispatchContext: React.Context<(action: any, callback?: () => void) => void>;
  reducer: Reducer<S>;
  initialState: S;
  persistKey: string;
}

/**
 * This function give you an store, use that in your components which want to
 * connect to store and provider
 */
export const createStore = <T extends { [x: string]: any }>({
  reducer,
  initialState,
  mapStateToPersist,
  Storage = window.localStorage,
  persistKey,
}: {
  reducer: Reducer<T>;
  initialState: T;
  Storage: any;
  persistKey: string;
  mapStateToPersist?: (state: T) => Partial<T>;
}): Store<T> => {
  const stateContext = createObserveContext(initialState);
  const dispatchContext = createContext<(action: any, callback?: any) => void>(
    () => {},
  );

  const store: PrivateStore<T> = {
    persistKey,
    dispatch: () => {},
    state: initialState,
    stateContext,
    dispatchContext,
    reducer,
    initialState,
    useState: (nextObserveState?: (keyof T)[]): T => {
      return useObserveContext(stateContext, nextObserveState);
    },
    useDispatch: () => useContext(dispatchContext),
    persist(state: T, action: { type: any }) {
      if (action.type !== INITIALIZE_STATE_FROM_STORAGE) {
        Storage.setItem(
          persistKey,
          JSON.stringify(mapStateToPersist ? mapStateToPersist(state) : state),
        );
      }
    },

    async setToState() {
      try {
        const storedState = await Storage.getItem(persistKey);

        if (!storedState) {
          return;
        }

        const stateObject = JSON.parse(storedState);

        const { dispatch } = store;
        const mappedState = mapStateToPersist
          ? mapStateToPersist(stateObject)
          : stateObject;

        dispatch({
          type: INITIALIZE_STATE_FROM_STORAGE,
          payload: mappedState,
        });

        /** Listening to events between tabs */
        window.addEventListener?.("storage", (event) => {
          if (event.key === persistKey && event.newValue !== event.oldValue) {
            dispatch({
              type: INITIALIZE_STATE_FROM_STORAGE,
              payload: event.newValue
                ? JSON.parse(event.newValue)
                : event.newValue,
            });
          }
        });
      } catch (error) {
        if (__DEV__) {
          console.error(error);
        }
      }
    },
  };

  return store as Store<T>;
};
