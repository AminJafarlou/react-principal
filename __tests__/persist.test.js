import renderer, { act } from "react-test-renderer";
import React, { useRef } from "react";
import { ExampleStoreTheme, changeTheme } from "../examples/store";
import { Provider, persisterCreator } from "../src";

const StorageMock = {
  data: {},
  setItem(key, data) {
    this.data[key] = data;
  },
  getItem(key) {
    return this.data[key];
  },
};

const persister = persisterCreator(StorageMock, "Theme");

const TextInputTester = function() {
  const { theme } = ExampleStoreTheme.useState(() => ({
    theme: true,
  }));
  return <p theme={theme} />;
};

let themeStoreRef;
const App = () => {
  themeStoreRef = useRef();
  return (
    <Provider
      ref={themeStoreRef}
      onStateDidChange={persister.persist}
      store={ExampleStoreTheme}
    >
      <TextInputTester />
    </Provider>
  );
};

let component;
let tree;
test("store: create theme store", async () => {
  component = renderer.create(<App />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  act(() => {
    const {
      dispatch,
      state: { theme },
    } = themeStoreRef.current;
    dispatch(changeTheme(theme === "light" ? "dark" : "light"));
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("store: create theme ", async () => {
  component = renderer.create(<App />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  await act(async () => {
    await persister.setToState(themeStoreRef);
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
