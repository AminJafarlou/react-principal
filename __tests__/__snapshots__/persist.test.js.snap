// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store: create theme  1`] = `
<p
  theme="light"
/>
`;

exports[`store: create theme  2`] = `
<p
  theme="dark"
/>
`;

exports[`store: create theme store 1`] = `
<p
  theme="light"
/>
`;

exports[`store: create theme store 2`] = `
<p
  theme="dark"
/>
`;
